Mathlib/Analysis/NormedSpace/Star/Matrix.lean

```diff
@@ -70,3 +89,151 @@ theorem entrywise_sup_norm_bound_of_unitary {U : Matrix n n 𝕜} (hU : U ∈ Ma
+    ‖A‖ = ‖(toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap A‖ := rfl
+
+lemma l2_op_nnnorm_def (A : Matrix m n 𝕜) :
+    ‖A‖₊ = ‖(toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap A‖₊ := rfl
+
+-- note: with only a type ascription in the left-hand side, Lean picks the wrong norm.
+lemma l2_op_norm_mulVec (A : Matrix m n 𝕜) (x : EuclideanSpace 𝕜 n) :
+    ‖(EuclideanSpace.equiv m 𝕜).symm <| A.mulVec x‖ ≤ ‖A‖ * ‖x‖ :=
+  toEuclideanLin (n := n) (m := m) (𝕜 := 𝕜) |>.trans toContinuousLinearMap A |>.le_op_norm x
+
+lemma l2_op_nnnorm_mulVec_le (A : Matrix m n 𝕜) (x : EuclideanSpace 𝕜 n) :
+    ‖(EuclideanSpace.equiv m 𝕜).symm <| A.mulVec x‖₊ ≤ ‖A‖₊ * ‖x‖₊ :=
+  A.l2_op_norm_mulVec x
+
+lemma l2_op_norm_mul {l : Type*} [Fintype l] [DecidableEq l] (A : Matrix m n 𝕜) (B : Matrix n l 𝕜) :
+    ‖A * B‖ ≤ ‖A‖ * ‖B‖ := by
+  simp only [l2_op_norm_def]
+  have := (toEuclideanLin (n := n) (m := m) (𝕜 := 𝕜) ≪≫ₗ toContinuousLinearMap) A
+    |>.op_norm_comp_le <| (toEuclideanLin (n := l) (m := n) (𝕜 := 𝕜) ≪≫ₗ toContinuousLinearMap) B
+  convert this
+  ext1 x
+  exact congr($(Matrix.toLin'_mul A B) x)
+
+lemma l2_op_nnnorm_mul {l : Type*} [Fintype l] [DecidableEq l] (A : Matrix m n 𝕜)
+    (B : Matrix n l 𝕜) : ‖A * B‖₊ ≤ ‖A‖₊ * ‖B‖₊ :=
+  l2_op_norm_mul A B
+
+/-- The normed algebra structure on `Matrix n n 𝕜` arising from the operator norm given by the
+identification with (continuous) linear endmorphisms of `EuclideanSpace 𝕜 n`. -/
+def instNormedSpaceL2Op : NormedSpace 𝕜 (Matrix m n 𝕜) where
+  norm_smul_le r x := by
+    rw [l2_op_norm_def, LinearEquiv.map_smul]
+    exact norm_smul_le r ((toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap x)
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instNormedSpaceL2Op
+
+/-- The normed ring structure on `Matrix n n 𝕜` arising from the operator norm given by the
+identification with (continuous) linear endmorphisms of `EuclideanSpace 𝕜 n`. -/
+def instNormedRingL2Op : NormedRing (Matrix n n 𝕜) where
+  dist_eq := normedRingAuxL2Op.dist_eq
+  norm_mul := normedRingAuxL2Op.norm_mul
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instNormedRingL2Op
+
+lemma cstar_norm_def (A : Matrix n n 𝕜) : ‖A‖ = ‖toEuclideanClm (n := n) (𝕜 := 𝕜) A‖ := rfl
+
+lemma cstar_nnnorm_def (A : Matrix n n 𝕜) : ‖A‖₊ = ‖toEuclideanClm (n := n) (𝕜 := 𝕜) A‖₊ := rfl
+
+/-- The normed algebra structure on `Matrix n n 𝕜` arising from the operator norm given by the
+identification with (continuous) linear endmorphisms of `EuclideanSpace 𝕜 n`. -/
+def instNormedAlgebraL2Op : NormedAlgebra 𝕜 (Matrix n n 𝕜) where
+  norm_smul_le := norm_smul_le
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instNormedAlgebraL2Op
+
+/-- The operator norm on `Matrix n n 𝕜` given by the identification with (continuous) linear
+endmorphisms of `EuclideanSpace 𝕜 n` makes it into a `L2OpRing`. -/
+lemma instCstarRing : CstarRing (Matrix n n 𝕜) where
```

Comment:
Is `NormOneClass` true too? Or is it automatic from `CstarRing`?

Reply:
Should be automatic with `Nontrivial (Matrix n n 𝕜)`, which holds whenever `n` is nonempty.